server/src/dependency.rs

```diff
@@ -0,0 +1,178 @@
+use {
+    move_binary_format::errors::PartialVMError,
+    moved::{
+        block::{BlockHash, BlockQueries, BlockRepository, MovedBlockHash},
+        move_execution::{BaseTokenAccounts, MovedBaseTokenAccounts},
+        receipt::{ReceiptQueries, ReceiptRepository},
+        transaction::{TransactionQueries, TransactionRepository},
+    },
+    moved_genesis::config::GenesisConfig,
+    moved_state::State,
+};
+
+#[cfg(feature = "storage")]
+pub type SharedStorage = &'static moved_storage::RocksDb;
+#[cfg(not(feature = "storage"))]
+pub type SharedStorage = moved::in_memory::SharedMemory;
+#[cfg(feature = "storage")]
+pub type ReceiptStorage = &'static moved_storage::RocksDb;
+#[cfg(not(feature = "storage"))]
+pub type ReceiptStorage = moved::receipt::ReceiptMemory;
+#[cfg(feature = "storage")]
+pub type StateQueries = moved::state_actor::InMemoryStateQueries;
+#[cfg(not(feature = "storage"))]
+pub type StateQueries = moved::state_actor::InMemoryStateQueries;
+
+pub fn block_hash() -> impl BlockHash + Send + Sync + 'static {
+    MovedBlockHash
+}
+
+pub fn base_token(
+    genesis_config: &GenesisConfig,
+) -> impl BaseTokenAccounts + Send + Sync + 'static {
+    MovedBaseTokenAccounts::new(genesis_config.treasury)
+}
+
+pub fn memory() -> SharedStorage {
+    #[cfg(feature = "storage")]
+    {
+        db()
+    }
+    #[cfg(not(feature = "storage"))]
+    {
+        moved::in_memory::SharedMemory::new()
+    }
+}
+
+pub fn block_repository() -> impl BlockRepository<Storage = SharedStorage> + Send + Sync + 'static {
+    #[cfg(feature = "storage")]
+    {
+        moved_storage::block::RocksDbBlockRepository
+    }
+    #[cfg(not(feature = "storage"))]
+    {
+        moved::block::InMemoryBlockRepository::new()
+    }
+}
+
+pub fn state() -> impl State<Err = PartialVMError> + Send + Sync + 'static {
+    #[cfg(feature = "storage")]
+    {
+        moved_storage::RocksDbState::new(std::sync::Arc::new(moved_storage::RocksEthTrieDb::new(
+            db(),
+        )))
+    }
+    #[cfg(not(feature = "storage"))]
+    {
+        moved_state::InMemoryState::new()
+    }
+}
+
+pub fn state_query(genesis_config: &GenesisConfig) -> StateQueries {
+    #[cfg(feature = "storage")]
+    {
+        moved::state_actor::InMemoryStateQueries::from_genesis(genesis_config.initial_state_root)
```

Comment:
I think we should save the state roots in the persistent storage too, so that when we make the `op-move` restartable it won't lose state root history.